### PR TITLE
Add merge and diff builtins

### PR DESCRIPTION
**Pull request by aaronlehmann**
Uses the new LLB "merge" and "diff" ops.

Example merge usage:

    fs alpine() {
            image "alpine"
    }

    fs busybox() {
            image "busybox"
    }

    fs merged() {
            busybox
            merge alpine
            breakpoint
    }

Example which uses merge and diff:

    fs alpine() {
            image "alpine"
    }

    fs busybox() {
            image "busybox"
    }

    fs workspace() {
            busybox
            run "echo this is a test > /root/file1"
            run "dd if=/dev/urandom bs=1024 count=100 of=/var/random"
            run "rm /bin/zcat"
    }

    fs differences() {
            workspace
            diff busybox
    }

    fs workspaceRebased() {
            alpine
            merge differences
    }

    fs default() {
            busybox
            run "ls" with option {
                    breakpoint
                    mount workspace "/workspace"
                    mount differences "/differences"
                    mount workspaceRebased "/rebased"
            }
    }

cc @sipsma

### CODE REVIEW

File: builtin/lookup.go [hunk 1]
```diff
@@ -1163,6 +1170,13 @@ option::copy includePatterns(variadic string pattern)
 # @return an option to copy files that don&#39;t match any pattern.
 option::copy excludePatterns(variadic string pattern)
 
+# Merges two input filesystems.
+#
+# @param input1 first filesystem to merge.
+# @param input2 second filesystem to merge.
+# @return a filesystem with input2 merged on top of input1
+fs merge(fs input1, fs input2)
```
Comment (coryb) [inline builtin/lookup.go:1178]: I wonder if this should be `fs merge(variadic fs input)` so we can `merge fs1 fs2 fs3 fs4`

**Review by coryb — COMMENTED** [review]
Looks promising.  Should probably add a few tests for this.

File: builtin/lookup.go (hunk 1 above)
Comment (aaronlehmann) [reply builtin/lookup.go]: Made it variadic and added a codegen test.

**Comment by aaronlehmann** [comment]
Example using merge and diff:

```
fs alpine() {
        image "alpine"
}

fs busybox() {
        image "busybox"
}

fs workspace() {
        busybox
        run "echo this is a test > /root/file1"
        run "dd if=/dev/urandom bs=1024 count=100 of=/var/random"
        run "rm /bin/zcat"
}

fs differences() {
        workspace
        diff busybox
}

fs workspaceRebased() {
        alpine
        merge differences
}

fs default() {
        busybox
        run "ls" with option {
                breakpoint
                mount workspace "/workspace"
                mount differences "/differences"
                mount workspaceRebased "/rebased"
        }
}
```

**Comment by aaronlehmann** [comment]
Split out the DiffOp part to https://github.com/aaronlehmann/hlb/pull/2. Now that MergeOp is available in upstream BuildKit, this PR is ready for review.

**Review by coryb — APPROVED** [review]
love it! :heart:

**Comment by aaronlehmann** [comment]
Pulled the `diff` builtin back into this PR, now that DiffOp is merged upstream in BuildKit.

File: codegen/codegen_test.go [hunk 2]
```diff
@@ -857,6 +857,22 @@ func TestCodeGen(t *testing.T) {
...
+			merge image("root1") image("root2")
+		}
+		`, "",
+		func(ctx context.Context, t *testing.T) solver.Request {
+			return Expect(t, llb.Merge([]llb.State{
+				llb.Image("alpine"),
+				llb.Image("root1"),
+				llb.Image("root2"),
+			}))
+		},
```
Comment (hinshun) [inline codegen/codegen_test.go:875]: Missing test for diff op, or is this not possible?

**Review by hinshun — CHANGES_REQUESTED** [review]

File: codegen/builtin_fs.go [hunk 3]
```diff
@@ -523,6 +524,50 @@ func (m Copy) Call(ctx context.Context, cln *client.Client, ret Register, opts O
...
+	fs.State = llb.Merge(states, SourceMap(ctx)...)
+
+	commitHistory(fs.Image, false, "MERGE %s %s", "/", "/")
+
+	return ret.Set(fs)
+}
+
+type Diff struct{}
+
+func (d Diff) Call(ctx context.Context, cln *client.Client, ret Register, opts Option, inputs ...Filesystem) error {
```
[the remark this answers is not in the document]
Comment (aaronlehmann) [reply codegen/builtin_fs.go]: This wasn't intentional - fixed

File: codegen/codegen_test.go (hunk 2 above)
Comment (aaronlehmann) [reply codegen/codegen_test.go:875]: Added a test case for `diff`

**Review by hinshun — APPROVED** [review]

**Pull request merged**